keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+#### Example
+
+```go
+import "k8s.io/component-base/kontext"
+
+func UpdateStatus(ctx context.Context, pods []*v1.Pod, newStatus bool) {
+	for _, pod := range pods {
+		podCtx := kontext.WithObject(ctx, pod)
+		updatePodStatus(podCtx, pod, newStatus)
+	}
+}
+
+func updatePodStatus(ctx context.Context, pod *v1.Pod, newStatus bool) {
+  logger := kontext.GetLogger(ctx).WithName("probe")
+  logger.V(6).Info("Updating ready status", "ready", newStatus)
+  [...]
+  logger.V(4).Info("Ready status updated", "ready", newStatus)
+}
+```
+
+### New logging formats
+
+Introduction of new logging API should be paired with new output format that can properly utilize structured logging. In order to simplify log ingestion, but still maintain similar developer experience we would like to propose two formats. First one for production use and second one for component development.
+
+#### Json format
+
+We would like to propose a JSON as a future default format in Kubernetes for both production and development use. Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```json
+{
+   "ts":"00:15:15.525108",
+   "msg":"Updating ready status",
+   "core/node":{
+      "name":"node-1"
+   },
+   "core/pod":{
+      "name":"nginx-1",
+      "namespace":"default"
+   },
+   "probe":{
+      "ready":false
+   }
+}
+```
+
+### Logging configuration
+
+We would like to introduce new logging configuration options shared by all kubernetes components. `LoggingConfig` structure should be implemented as part of `k8s.io/component-base` options and include common set of flags for logger initialization.
+* `--logging-format` flag should allow to pick between logging interface implementation and allow for backward compatibility. Setting this flag will select particular logger implementation from registry.
+* `-v` flag should allow to change verbosity of logging
+
+Any additional configuration needed by loggers should be provided by flags specific to its implementation. Default loggers should provide a basic customization, but it will not be encouraged to avoid feature bloat. Adding new loggers, changing logger behavior or extending its functionality should be done outside of Kubernetes by implementing logr API and registering it in code.
+
+We acknowledge that extension via code is not the best, but should be sufficient for now.
+
+Proposed flag `--logging-format`  values:
+* `legacy` for old logging format
```

Comment:
it's not obvious to me how you will have things like "Object" serialize reasonably in a human targeted log like we have today.

Reply:
Problem we need to tackle here is inconsistency in referencing objects. Usually it's the name alone e.g. `Logf("Error detected on %v : %v !", pod.Name, err)`, sometimes name and namespace e.g. `Infof("Metrics not yet available for pod %s/%s, age: %s", pod.Namespace, pod.Name, age.String())`.

We are proposing to have standard format "<namespace>/<name>" and provide function that would do this format automatically.